crypto-pgpainless/src/main/kotlin/dev/msfjarvis/aps/crypto/PGPKeyManager.kt

```diff
@@ -50,63 +63,105 @@ public class PGPKeyManager(
       }
     }
 
-  override suspend fun getKeyById(id: String): Result<PGPKeyPair, Throwable> =
+  override suspend fun getKeyById(id: String): Result<Key, Throwable> =
     withContext(dispatcher) {
       runCatching {
         if (!keyDirExists()) throw KeyManagerException.KeyDirectoryUnavailableException
-        val keys = keyDir.listFiles()
-        if (keys.isNullOrEmpty()) throw KeyManagerException.NoKeysAvailableException
+        val keyFiles = keyDir.listFiles()
+        if (keyFiles.isNullOrEmpty()) throw KeyManagerException.NoKeysAvailableException
+
+        val keys = keyFiles.map { file -> Key(file.readBytes()) }
+        // Try to parse the key ID as an email
+        val selector = SelectUserId.byEmail(id)
+        val userIdMatch =
+          keys.map { key -> key to tryParseKeyring(key) }.firstOrNull { (_, keyRing) ->
+            selector.firstMatch(keyRing) != null
+          }
+
+        if (userIdMatch != null) {
+          return@runCatching userIdMatch.first
+        }
+
+        val keyIdMatch =
+          keys.map { key -> key to tryGetId(key) }.firstOrNull { (_, keyId) ->
+            keyId == id || keyId == "0x$id"
+          }
 
-        for (keyFile in keys) {
-          val secretKeyRing = PGPainless.readKeyRing().secretKeyRing(keyFile.inputStream())
-          val secretKey = secretKeyRing.secretKey
-          val keyPair = PGPKeyPair(secretKey)
-          if (keyPair.getKeyId() == id) return@runCatching keyPair
+        if (keyIdMatch != null) {
+          return@runCatching keyIdMatch.first
         }
 
         throw KeyManagerException.KeyNotFoundException(id)
       }
     }
 
-  override suspend fun getAllKeys(): Result<List<PGPKeyPair>, Throwable> =
+  override suspend fun getAllKeys(): Result<List<Key>, Throwable> =
     withContext(dispatcher) {
       runCatching {
         if (!keyDirExists()) throw KeyManagerException.KeyDirectoryUnavailableException
         val keys = keyDir.listFiles()
         if (keys.isNullOrEmpty()) return@runCatching listOf()
```

Comment:
could be renamed to `keyFiles` to maintain consistency with `getKeyById`

Reply:
Done